File: claude/mcp.test.mjs
// SPDX-License-Identifier: MIT
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { PassThrough, Writable } from 'node:stream';
import { serve, hiddenTool } from './mcp.mjs';

function harness(owner) {
  const input=new PassThrough(), output=new PassThrough(), pending=[], received=[];
  let buffer='';output.on('data',b=>{buffer+=b;for(let n;(n=buffer.indexOf('\n'))>=0;){const x=JSON.parse(buffer.slice(0,n));buffer=buffer.slice(n+1);if(pending.length)pending.shift()(x);else received.push(x);}});
  const service=serve(owner,input,output);
  return {input,output,service,send:x=>input.write(JSON.stringify(x)+'\n'),next:()=>received.length?Promise.resolve(received.shift()):new Promise(r=>pending.push(r))};
}
const request=(id,method,params={})=>({jsonrpc:'2.0',id,method,params});
test('initialize, ping and catalog; hidden native-shaped call excluded from public actions',async()=>{
  const reports=[];const h=harness({report:async e=>reports.push(e),end(){}});
  h.send(request(1,'initialize',{protocolVersion:'2025-11-25'})); assert.equal((await h.next()).result.protocolVersion,'2025-11-25');
  h.send(request(2,'tools/list')); assert.deepEqual((await h.next()).result.tools.map(t=>t.name),['sessionbus']);
  const event={hook_event_name:'Stop',session_id:'actual-native',session_title:''};
  h.send(request(3,'tools/call',{name:hiddenTool,arguments:event})); assert.ok((await h.next()).result.content);assert.deepEqual(reports,[event]);
  h.send(request(4,'ping'));assert.deepEqual((await h.next()).result,{});h.service.close();
});
test('pending tool cannot block report, notification or EOF; late answer never writes after close',async()=>{
  let finish, markEnd; const ended=new Promise(r=>markEnd=r); const reports=[];
  const h=harness({action:()=>new Promise(r=>finish=r),report:async e=>reports.push(e),end(){markEnd();}});
  h.send(request(1,'tools/call',{name:'sessionbus',arguments:{action:'list',arguments:{}}}));
  h.send({jsonrpc:'2.0',method:'notifications/initialized'});
  h.send(request(2,'tools/call',{name:hiddenTool,arguments:{hook_event_name:'SessionEnd',session_id:'id'}}));
  assert.equal((await h.next()).id,2);assert.equal(reports.length,1);
  h.input.end();await ended;finish({sessions:[]});await Promise.resolve();h.output.destroy();
});
test('malformed frames, unknown methods, invalid public arguments and report errors are protocol replies',async()=>{
  const h=harness({end(){},report(){throw Error('unusable native identity report');}});
  h.input.write('{bad\n');assert.equal((await h.next()).error.code,-32700);
  h.send(request(1,'unknown'));assert.equal((await h.next()).error.code,-32601);
  h.send(request(2,'tools/call',{name:hiddenTool,arguments:{}}));assert.equal((await h.next()).result.isError,true);
  h.send(request(3,'tools/call',{name:'sessionbus',arguments:{action:hiddenTool,arguments:{}}}));assert.equal((await h.next()).result.isError,true);
  h.service.close();
});
test('output failure terminates owner and closes report processing',async()=>{
  let closed;const done=new Promise(r=>closed=r),input=new PassThrough();
  const output=new Writable({write(_b,_e,cb){cb(new Error('controlled output failure'));}});
  serve({end:closed},input,output);input.write(JSON.stringify(request(1,'ping'))+'\n');await done;input.destroy();
});
test('published stdio EOF closes actual exported Connection and pending Caller request',async t=>{
  const {createServer}=await import('node:net'); const {mkdtempSync,rmSync}=await import('node:fs');
  const {tmpdir}=await import('node:os'); const {join}=await import('node:path'); const {once}=await import('node:events');
  const {Connection}=await import('@sessionbus/kit'); const {Owner}=await import('./owner.mjs');
  const dir=mkdtempSync(join(tmpdir(),'claude-mcp-')),path=join(dir,'s');let peer, called;
  const pending=new Promise(r=>called=r);
  const server=createServer(s=>{peer=new Connection(s,false,req=>{
    if(req.method==='session.hello') void peer.result(req,{}); else called(req);
  });});server.listen(path);await once(server,'listening');
  const owner=new Owner({env:{SESSIONBUS_SOCKET:path}}),h=harness(owner);
  t.after(async()=>{h.service.close();peer?.close();await new Promise(r=>server.close(r));rmSync(dir,{recursive:true,force:true});});
  h.send(request(1,'tools/call',{name:hiddenTool,arguments:{hook_event_name:'Stop',session_id:'reported'}}));
  await h.next();assert.equal(owner.admitted.session_id,'reported');
  h.send(request(2,'tools/call',{name:'sessionbus',arguments:{action:'list',arguments:{}}}));
  assert.equal((await pending).method,'session.list');const connection=owner.connection;
  h.input.end();await connection.done;await peer.done;
  assert.equal(owner.ended,true);assert.equal(connection.pending.size,0);assert.equal(owner.admitted,null);
});
test('MCP cancellation targets only the pending public request and removes it on completion',async()=>{
  const calls=[];let finish;
  const h=harness({end(){},action(_a,_v,signal){calls.push(signal);return new Promise((resolve,reject)=>{finish=resolve;signal.addEventListener('abort',()=>reject(signal.reason),{once:true});});}});
  h.send(request(10,'tools/call',{name:'sessionbus',arguments:{action:'list',arguments:{}}}));
  h.send({jsonrpc:'2.0',method:'notifications/cancelled',params:{requestId:99}});assert.equal(calls[0].aborted,false);
  h.send({jsonrpc:'2.0',method:'notifications/cancelled',params:{requestId:10}});assert.equal(calls[0].aborted,true);
  h.send(request(11,'ping'));assert.equal((await h.next()).id,11);finish({sessions:[]});await Promise.resolve();
  h.send(request(12,'tools/call',{name:'sessionbus',arguments:{action:'list',arguments:{}}}));finish({sessions:[]});
  assert.equal((await h.next()).id,12);
  h.send({jsonrpc:'2.0',method:'notifications/cancelled',params:{requestId:12}});assert.equal(calls[1].aborted,false);
  h.service.close();
});

for (const collect of ['status','wait']) test(`cancelled MCP wait leaves its result collectible by ${collect}`,async t=>{
  const {createServer}=await import('node:net');const {mkdtempSync,rmSync}=await import('node:fs');
  const {tmpdir}=await import('node:os');const {join}=await import('node:path');const {once}=await import('node:events');
  const {Connection}=await import('@sessionbus/kit');const {Owner}=await import('./owner.mjs');
  const dir=mkdtempSync(join(tmpdir(),'claude-wait-')),path=join(dir,'s');const queue=[],readers=[];let peer;
  const server=createServer(stream=>{peer=new Connection(stream,false,r=>{
    if(r.method==='session.hello') void peer.result(r,{});
    else if(readers.length) readers.shift()(r);else queue.push(r);
  });});server.listen(path);await once(server,'listening');
  const next=()=>queue.length?Promise.resolve(queue.shift()):new Promise(r=>readers.push(r));
  const owner=new Owner({env:{SESSIONBUS_SOCKET:path}}),h=harness(owner);
  t.after(async()=>{h.service.close();peer?.close();await new Promise(r=>server.close(r));rmSync(dir,{recursive:true,force:true});});
  const call=(id,action,args)=>h.send(request(id,'tools/call',{name:'sessionbus',arguments:{action,arguments:args}}));
  const value=frame=>JSON.parse(frame.result.content[0].text);
  h.send(request(1,'tools/call',{name:hiddenTool,arguments:{hook_event_name:'Stop',session_id:'reported'}}));await h.next();
  call(2,'start',{session_id:'daemon-lane',input:'work'});const {turn_id}=value(await h.next());const run=await next();assert.equal(run.method,'turn.run');
  call(3,'wait',{turn_id});h.send(request(4,'ping'));assert.equal((await h.next()).id,4);
  h.send({jsonrpc:'2.0',method:'notifications/cancelled',params:{requestId:3}});
  h.send(request(5,'ping'));assert.equal((await h.next()).id,5);
  // Ordered actual Connection replies provide a public completion barrier.
  await peer.result(run,{outcome:'completed',result:'keep-this-result'});
  call(6,'list',{});const list=await next();assert.equal(list.method,'session.list');await peer.result(list,{sessions:[]});assert.equal((await h.next()).id,6);
  call(7,collect,{turn_id});const collected=await h.next();assert.equal(collected.id,7);
  assert.deepEqual(value(collected),{turn_id,session_id:'daemon-lane',state:'done',result:{outcome:'completed',result:'keep-this-result'}});
  call(8,'status',{turn_id});const consumed=await h.next();assert.equal(consumed.id,8);assert.equal(consumed.result.isError,true);assert.match(consumed.result.content[0].text,/unknown_turn/);
});

test('fulfilled consuming operation keeps its MCP response when cancellation arrives afterwards',async()=>{
  const {Caller,Connection}=await import('@sessionbus/kit');
  const {Duplex}=await import('node:stream');
  let clientStream,serverStream;
  clientStream=new Duplex({read(){},write(b,_e,cb){serverStream.push(b);cb();}});
  serverStream=new Duplex({read(){},write(b,_e,cb){clientStream.push(b);cb();}});
  let complete;const requested=new Promise(r=>complete=r);
  const server=new Connection(serverStream,false,r=>complete(r));const client=new Connection(clientStream,true);
  const caller=new Caller(client);let h;
  const owner={end(){caller.disconnected();client.close();server.close();},action:async(action,args,signal)=>{
    const value=await caller.action(action,args,signal);
    if(action==='wait') h.send({jsonrpc:'2.0',method:'notifications/cancelled',params:{requestId:2}});
    return value;
  }};
  h=harness(owner);
  h.send(request(1,'tools/call',{name:'sessionbus',arguments:{action:'start',arguments:{session_id:'lane',input:'work'}}}));
  const {turn_id}=JSON.parse((await h.next()).result.content[0].text);const run=await requested;
  h.send(request(2,'tools/call',{name:'sessionbus',arguments:{action:'wait',arguments:{turn_id}}}));
  await server.result(run,{outcome:'completed',result:'completion-won'});
  const frame=await h.next();assert.equal(frame.id,2);assert.equal(JSON.parse(frame.result.content[0].text).result.result,'completion-won');
  assert.throws(()=>caller.status({turn_id}),/unknown_turn/);h.service.close();
});

File: wrappers/claude/interactive/mcp_test.go
// SPDX-License-Identifier: MIT

package interactive

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
	"testing"

	kit "github.com/antst/sessionbus/bus/sdk/go"
	"github.com/antst/sessionbus/bus/sdk/go/protocol"
)

type mcpHarness struct {
	input  *io.PipeWriter
	frames chan map[string]json.RawMessage
	done   chan struct{}
}

func newMCP(t *testing.T, o MCPOwner) *mcpHarness {
	t.Helper()
	in, w := io.Pipe()
	out, r := io.Pipe()
	h := &mcpHarness{input: w, frames: make(chan map[string]json.RawMessage, 32), done: make(chan struct{})}
	go func() { defer close(h.done); _ = Serve(o, in, r); _ = r.Close() }()
	go func() {
		defer close(h.frames)
		d := json.NewDecoder(out)
		for {
			var f map[string]json.RawMessage
			if d.Decode(&f) != nil {
				return
			}
			h.frames <- f
		}
	}()
	t.Cleanup(func() { _ = w.Close(); <-h.done; _ = out.Close() })
	return h
}
func (h *mcpHarness) send(t *testing.T, frame any) {
	t.Helper()
	b, _ := json.Marshal(frame)
	if _, err := h.input.Write(append(b, '\n')); err != nil {
		t.Fatal(err)
	}
}
func (h *mcpHarness) call(t *testing.T, id int, action string, args any) {
	t.Helper()
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": id, "method": "tools/call", "params": map[string]any{"name": "sessionbus", "arguments": map[string]any{"action": action, "arguments": args}}})
}
func (h *mcpHarness) next(t *testing.T) map[string]json.RawMessage {
	t.Helper()
	f, ok := <-h.frames
	if !ok {
		t.Fatal("MCP closed")
	}
	return f
}
func (h *mcpHarness) cancel(t *testing.T, id int) {
	h.send(t, map[string]any{"jsonrpc": "2.0", "method": "notifications/cancelled", "params": map[string]any{"requestId": id}})
}
func value(t *testing.T, f map[string]json.RawMessage) json.RawMessage {
	t.Helper()
	var result struct {
		Content []struct{ Text string }
		IsError bool
	}
	if err := json.Unmarshal(f["result"], &result); err != nil || len(result.Content) != 1 {
		t.Fatalf("bad result %s", f["result"])
	}
	return json.RawMessage(result.Content[0].Text)
}

type observedOwner struct {
	*Owner
	waitEntered  chan struct{}
	waitReturned chan error
	listReturned chan error
	collected    chan struct{}
	release      chan struct{}
	once         sync.Once
}

func (o *observedOwner) Action(ctx context.Context, action string, args json.RawMessage) (json.RawMessage, error) {
	if action == "wait" && o.waitEntered != nil {
		o.once.Do(func() { close(o.waitEntered) })
	}
	raw, err := o.Owner.Action(ctx, action, args)
	if action == "list" && o.listReturned != nil {
		o.listReturned <- err
	}
	if action == "wait" && o.waitReturned != nil {
		o.waitReturned <- err
	}
	if action == "wait" && err == nil && o.collected != nil {
		close(o.collected)
		<-o.release
	}
	return raw, err
}

func startMCPRun(t *testing.T, h *mcpHarness, w *wire) kit.RunRef {
	t.Helper()
	h.call(t, 1, "start", map[string]string{"session_id": "lane", "input": "one"})
	request := w.next(t)
	if request.Method != "turn.start" {
		t.Fatal(request.Method)
	}
	ref := kit.RunRef{SessionID: "lane", RunID: "g/1"}
	w.reply(t, request, ref)
	var returned kit.RunRef
	if json.Unmarshal(value(t, h.next(t)), &returned) != nil || returned != ref {
		t.Fatal(returned)
	}
	return ref
}

func TestCancelledMCPWaitPreservesActualCallerResult(t *testing.T) {
	for _, collector := range []string{"status", "wait"} {
		t.Run(collector, func(t *testing.T) {
			o, wires := testOwner(t)
			w := published(t, o, wires)
			observed := &observedOwner{Owner: o, waitReturned: make(chan error, 2)}
			h := newMCP(t, observed)
			ref := startMCPRun(t, h, w)
			h.call(t, 2, "wait", ref)
			abandoned := w.next(t)
			if abandoned.Method != "turn.wait" {
				t.Fatal(abandoned.Method)
			}
			h.cancel(t, 2)
			if err := <-observed.waitReturned; !errors.Is(err, context.Canceled) {
				t.Fatal(err)
			}
			terminal := kit.RunStatus{SessionID: ref.SessionID, RunID: ref.RunID, State: "done", Result: &kit.TurnResult{Outcome: "completed", Result: "retained"}}
			w.reply(t, abandoned, terminal)
			// Collection now routes to the worker; no local handle is consumed.
			for _, callID := range []int{3, 4} {
				h.call(t, callID, collector, ref)
				read := w.next(t)
				if read.Method != "turn."+collector {
					t.Fatal(read.Method)
				}
				w.reply(t, read, terminal)
				f := h.next(t)
				var status kit.RunStatus
				if string(f["id"]) != fmt.Sprint(callID) || json.Unmarshal(value(t, f), &status) != nil || status.Result == nil || status.Result.Result != "retained" {
					t.Fatal(f)
				}
			}
			h.call(t, 5, "ack", ref)
			ack := w.next(t)
			if ack.Method != "turn.ack" {
				t.Fatal(ack.Method)
			}
			w.reply(t, ack, struct{}{})
			if string(h.next(t)["id"]) != "5" {
				t.Fatal("missing explicit ack result")
			}
		})
	}
}

func TestFulfilledWaitKeepsResponseAfterLateCancellation(t *testing.T) {
	o, wires := testOwner(t)
	w := published(t, o, wires)
	observed := &observedOwner{Owner: o, collected: make(chan struct{}), release: make(chan struct{})}
	h := newMCP(t, observed)
	ref := startMCPRun(t, h, w)
	h.call(t, 2, "wait", ref)
	read := w.next(t)
	w.reply(t, read, kit.RunStatus{SessionID: ref.SessionID, RunID: ref.RunID, State: "done", Result: &kit.TurnResult{Outcome: "completed", Result: "won"}})
	<-observed.collected
	h.cancel(t, 2)
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 9, "method": "ping"})
	if string(h.next(t)["id"]) != "9" {
		t.Fatal("cancellation barrier")
	}
	close(observed.release)
	f := h.next(t)
	var result kit.RunStatus
	_ = json.Unmarshal(value(t, f), &result)
	if string(f["id"]) != "2" || result.Result == nil || result.Result.Result != "won" {
		t.Fatal("fulfilled result dropped")
	}
}

func TestMCPProtocolHiddenReportAndEOF(t *testing.T) {
	o, wires := testOwner(t)
	h := newMCP(t, o)
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 1, "method": "tools/list"})
	f := h.next(t)
	var list struct{ Tools []struct{ Name string } }
	_ = json.Unmarshal(f["result"], &list)
	if len(list.Tools) != 1 || list.Tools[0].Name != "sessionbus" {
		t.Fatal(list)
	}
	h.send(t, map[string]any{"jsonrpc": "2.0", "method": "native_identity_event", "params": map[string]string{"session_id": "fake"}})
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 2, "method": "tools/call", "params": map[string]any{"name": HiddenTool, "arguments": map[string]string{"hook_event_name": "Stop", "session_id": "native"}}})
	w := <-wires
	hello := w.next(t)
	w.reply(t, hello, map[string]any{})
	_ = h.next(t)
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 3, "method": "unknown"})
	if string(h.next(t)["error"]) == "" {
		t.Fatal("missing protocol error")
	}
	_ = h.input.Close()
	<-h.done
	o.mu.Lock()
	defer o.mu.Unlock()
	if !o.ended || o.connection != nil {
		t.Fatal("EOF left integration alive")
	}
}

type failedOutput struct{}

func (failedOutput) Write([]byte) (int, error) { return 0, io.ErrClosedPipe }

func TestOutputFailureEndsOwnerAndUnblocksInput(t *testing.T) {
	o, _ := testOwner(t)
	r, w := io.Pipe()
	defer w.Close()
	done := make(chan struct{})
	go func() { _ = Serve(o, r, failedOutput{}); close(done) }()
	_, _ = io.WriteString(w, "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"ping\"}\n")
	<-done
	o.mu.Lock()
	defer o.mu.Unlock()
	if !o.ended {
		t.Fatal("output failure left owner active")
	}
}

func TestMCPRejectsNullVersionAndMalformedFrames(t *testing.T) {
	o, _ := testOwner(t)
	h := newMCP(t, o)
	for _, raw := range []string{
		`{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":null}}`,
		`{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":4}}`,
		`{"jsonrpc":"2.0","id":1,"method":"ping","params":[]}`,
		`{"jsonrpc":"2.0","id":null,"method":"ping"}`,
		`null`, `{`,
	} {
		if _, err := h.input.Write([]byte(raw + "\n")); err != nil {
			t.Fatal(err)
		}
		if f := h.next(t); len(f["error"]) == 0 {
			t.Fatalf("accepted %s", raw)
		}
	}
	for _, raw := range []string{
		`{"hook_event_name":"Stop","session_id":"id","agent_id":null}`,
		`{"hook_event_name":"Stop","session_id":"id","session_title":null}`,
		`{"hook_event_name":"Stop","session_id":"id","agent_id":"agent"}`,
		`{"hook_event_name":"SessionStart","session_id":"id"}`,
	} {
		if _, err := o.BeginReport(json.RawMessage(raw)); err == nil {
			t.Fatalf("accepted report %s", raw)
		}
	}
}

func TestPublishedMCPWithPendingCallerSettlesOnEOF(t *testing.T) {
	o, wires := testOwner(t)
	observed := &observedOwner{Owner: o, listReturned: make(chan error, 1)}
	h := newMCP(t, observed)
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 1, "method": "tools/call", "params": map[string]any{"name": HiddenTool, "arguments": map[string]string{"hook_event_name": "UserPromptSubmit", "session_id": "native-id", "session_title": "name"}}})
	w := <-wires
	hello := w.next(t)
	if hello.Method != "session.hello" {
		t.Fatal(hello.Method)
	}
	w.reply(t, hello, map[string]any{})
	if string(h.next(t)["id"]) != "1" {
		t.Fatal("publication response")
	}
	h.call(t, 2, "list", map[string]any{})
	request := w.next(t)
	if request.Method != "session.list" {
		t.Fatal(request.Method)
	}
	// This real Caller request remains unanswered; EOF alone must settle it.
	if err := h.input.Close(); err != nil {
		t.Fatal(err)
	}
	<-h.done
	if err := <-observed.listReturned; err == nil {
		t.Fatal("pending request falsely succeeded")
	}
	for f := range h.frames {
		t.Fatalf("response after EOF: %s", f["id"])
	}
	for f := range w.frames {
		t.Fatalf("new wire frame after EOF: %s", f.Method)
	}
	late, err := protocol.ResultBytes(request.ID, request.Method, map[string]any{"sessions": []any{}})
	if err != nil {
		t.Fatal(err)
	}
	if _, err = w.fd.Write(late); err == nil {
		t.Fatal("EOF left public Connection open")
	}
	if _, err = o.BeginReport(json.RawMessage(`{"hook_event_name":"Stop","session_id":"later"}`)); err == nil {
		t.Fatal("report revived ended owner")
	}
	select {
	case <-wires:
		t.Fatal("reconnected after EOF")
	default:
	}
}

func TestSubmittedMCPAckKeepsResponseAfterCancellation(t *testing.T) {
	o, wires := testOwner(t)
	w := published(t, o, wires)
	h := newMCP(t, o)
	h.call(t, 1, "ack", kit.RunRef{SessionID: "lane", RunID: "g/1"})
	ack := w.next(t)
	if ack.Method != "turn.ack" {
		t.Fatal(ack.Method)
	}
	h.cancel(t, 1)
	h.send(t, map[string]any{"jsonrpc": "2.0", "id": 2, "method": "ping"})
	if string(h.next(t)["id"]) != "2" {
		t.Fatal("ack returned before its wire result")
	}
	w.reply(t, ack, struct{}{})
	result := h.next(t)
	if string(result["id"]) != "1" || string(value(t, result)) != "{}" {
		t.Fatal("submitted acknowledgement lost its successful response")
	}
}
